crates/miden-objects/src/account/component/storage/type_registry.rs

```diff
@@ -0,0 +1,645 @@
+
+use miden_core::utils::{ByteReader, ByteWriter, Deserializable, Serializable};
+use miden_core::{Felt, Word};
+use miden_crypto::dsa::{ecdsa_k256_keccak, rpo_falcon512};
+use miden_processor::DeserializationError;
+use thiserror::Error;
+
+use crate::asset::TokenSymbol;
+use crate::utils::sync::LazyLock;
+
+/// A global registry for schema type converters.
+///
+/// It is used during component instantiation to convert init-provided values (typically provided
+/// as strings) into their respective storage values.
+pub static SCHEMA_TYPE_REGISTRY: LazyLock<SchemaTypeRegistry> = LazyLock::new(|| {
+    let mut registry = SchemaTypeRegistry::new();
+    registry.register_felt_type::<Void>();
+    registry.register_felt_type::<u8>();
+    registry.register_felt_type::<u16>();
+    registry.register_felt_type::<u32>();
+    registry.register_felt_type::<Felt>();
+    registry.register_felt_type::<TokenSymbol>();
+    registry.register_word_type::<Word>();
+    registry.register_word_type::<rpo_falcon512::PublicKey>();
+    registry.register_word_type::<ecdsa_k256_keccak::PublicKey>();
+    registry
+});
+
+// SCHEMA TYPE ERROR
+// ================================================================================================
+
+/// Errors that can occur when parsing or converting schema types.
+///
+/// This enum covers various failure cases including parsing errors, conversion errors,
+/// unsupported conversions, and cases where a required type is not found in the registry.
+#[derive(Debug, Error)]
+pub enum SchemaTypeError {
+    #[error("conversion error: {0}")]
+    ConversionError(String),
+    #[error("felt type ` {0}` not found in the type registry")]
+    FeltTypeNotFound(SchemaTypeIdentifier),
+    #[error("invalid type name `{0}`: {1}")]
+    InvalidTypeName(String, String),
+    #[error("failed to parse input `{input}` as `{schema_type}`")]
+    ParseError {
+        input: String,
+        schema_type: SchemaTypeIdentifier,
+        source: Box<dyn Error + Send + Sync + 'static>,
+    },
+    #[error("word type ` {0}` not found in the type registry")]
+    WordTypeNotFound(SchemaTypeIdentifier),
+}
+
+impl SchemaTypeError {
+    /// Creates a [`SchemaTypeError::ParseError`].
+    pub fn parse(
+        input: impl Into<String>,
+        schema_type: SchemaTypeIdentifier,
+        source: impl Error + Send + Sync + 'static,
+    ) -> Self {
+        SchemaTypeError::ParseError {
+            input: input.into(),
+            schema_type,
+            source: Box::new(source),
+        }
+    }
+}
+
+// SCHEMA TYPE
+// ================================================================================================
+
+/// A newtype wrapper around a `String`, representing a schema type identifier.
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+#[cfg_attr(feature = "std", derive(::serde::Deserialize, ::serde::Serialize))]
+#[cfg_attr(feature = "std", serde(transparent))]
+pub struct SchemaTypeIdentifier(String);
```

Comment:
nit: I'd call this just `SchemaTypeId` for brevity.

Also, I'd update the comment on line 77 to explain what valid identifiers look like.